### PR TITLE
Implemented Hash, HashSequenceStart, SequenceUpdate, and SequenceComplete

### PR DESCRIPTION
**Pull request by matt-tsai**
Wrote implemented and wrote tests for the following functions: Hash, HashSequenceStart, SequenceUpdate, and SequenceComplete.

### CODE REVIEW

File: direct/tpm2/tpm2.go [hunk 1]
```diff
@@ -387,6 +387,41 @@ type CreateLoadedResponse struct {
...
 
+// Hash is the input to TPM2_Hash.
+// See definition in Part 3, Commands, section 15.4
+type Hash struct {
+	//data to be hashed
+	Data tpm2b.MaxBuffer
+	// algorithm for the hash being computed - shall not be TPM_ALH_NULL
+	HashAlg tpmi.AlgHash
+	// hierarchy to use for the tickey (TPM_RH_NULL_allowed)
+	Hierarchy tpmi.RHHierarchy
```
Comment (josephlr) [inline direct/tpm2/tpm2.go:398]: If this member is omitted, will TPM_RH_NULL be used by default?

Comment (josephlr) [reply direct/tpm2/tpm2.go:398]: Regardless, we should have a test that using a Null Hierarchy with `TPM2_Hash` and `TPM2_SequenceComplete` works as expected.

Comment (matt-tsai) [reply direct/tpm2/tpm2.go:398]: > If this member is omitted, will TPM_RH_NULL be used by default?

Yes, it has now been added. The tests are being made now.

File: direct/tpm2/hash_sequence_hash_test.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+func TestHashSequence(t *testing.T) {
+	thetpm, err := simulator.OpenSimulator()
+	if err != nil {
+		t.Fatalf("could not connect to TPM simulator: %v", err)
+	}
+	defer thetpm.Close()
+
+	run := func(t *testing.T, bufferSize int, password string, hierarchy tpm.Handle, thetpm transport.TPM) {
+		maxDigestBuffer := 1024
```
Comment (chrisfenner) [inline direct/tpm2/hash_sequence_hash_test.go:83]: a TPM2B_MAX_BUFFER has maximum size "MAX_DIGEST" which can be read from the TPM as a "PT_MAX_DIGEST" and is always at least 1024 (so this is safe)

File: direct/tpm2/hash_sequence_hash_test.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+			t.Fatalf("SequenceComplete failed: %v", err)
+		}
+
+		gotDigest := rspSC.Result.Buffer
+
+		if !bytes.Equal(gotDigest, wantDigest[:]) {
+			t.Errorf("The resulting digest %x, is not expected %x", gotDigest, wantDigest)
+		}
+	}
+	//  t *testing.T, bufferSize int, password string, hierarchy tpm.Handle, thetpm transport.TPM
```
Comment (chrisfenner) [inline direct/tpm2/hash_sequence_hash_test.go:144]: Nit: spare comment

File: direct/tpm2/tpm2.go [hunk 4]
```diff
@@ -387,6 +387,41 @@ type CreateLoadedResponse struct {
...
 func (*CreateLoadedResponse) Response() tpm.CC { return tpm.CCCreateLoaded }
 
+// Hash is the input to TPM2_Hash.
+// See definition in Part 3, Commands, section 15.4
+type Hash struct {
+	//data to be hashed
+	Data tpm2b.MaxBuffer
+	// algorithm for the hash being computed - shall not be TPM_ALH_NULL
+	HashAlg tpmi.AlgHash
+	// hierarchy to use for the tickey (TPM_RH_NULL_allowed)
```
Comment (chrisfenner) [inline direct/tpm2/tpm2.go:397]: nit: ticket

**Review by chrisfenner — APPROVED** [review]
LGTM - just a couple super minor nits but ready to merge when fixed!

File: direct/tpm2/hash_sequence_hash_test.go (hunk 3 above)
Comment (matt-tsai) [reply direct/tpm2/hash_sequence_hash_test.go:144]: Resolved

File: direct/tpm2/tpm2.go (hunk 4 above)
Comment (matt-tsai) [reply direct/tpm2/tpm2.go:397]: Resolved

File: direct/tpm2/hash_sequence_hash_test.go (hunk 2 above)
Comment (matt-tsai) [reply direct/tpm2/hash_sequence_hash_test.go:83]: Acknowledged!

**Comment by alexmwu** [comment]
Double checked Chris's comments were addressed, thanks for this!

**Pull request merged**